crates/wasmtime/src/runtime/vm/table.rs

```diff
@@ -701,29 +726,60 @@ impl Table {
         self.element_type().matches(val)
     }
 
-    fn funcrefs(&self) -> &[TaggedFuncRef] {
+    fn funcrefs(&self) -> (&[TaggedFuncRef], bool) {
```

Comment:
What do you think about perhaps returning `enum Funcrefs<'a> { Tagged(&'a [TaggedFuncRef]), Untagged(&'a [*mut VMFuncRef]) }` here? That feels a bit more "type safe" if a bit wordier and reflects how this is a boolean decision at runtime as well

Reply:
I like this idea, but while testing it out, I found so many more things I want to change about how the lazy-tagging invariants are maintained in these modules that I don't think I want to change any of them in this PR. :sweat_smile: A few:
- Getting a table reference should return enough information that the caller can initialize any elements it needs to, instead of passing in a range of indexes to init.
- `Table::get` and table copies should take `&mut src` and init the elements they need in-place; the caller always has a mutable pointer.
- Therefore, there's no need for `funcrefs()`, only `funcrefs_mut()`.
- `StaticFuncTable` should maybe put the current size in the `data` slice and the capacity in a separate field; it's currently the other way around.
- `funcrefs_mut()` is shorter and easier to reason about if it first gets `elements.as_mut_slice()`/`unsafe { data.as_mut() }` depending on whether the table is static or dynamic, and then afterwards does the unsafe `slice::from_raw_parts_mut`. I mixed up `*mut [T]` with `*mut [*mut [T]]` at one point and was afraid the `.cast()` would just silently "fix" it for me
- I'm imagining something resembling `MaybeUninit` as an interface for tracking at the type level whether it's okay to read a range of the table

Reply:
That all sounds great to me yeah, and seems fine to leave this as-is returning a bool in the meantime